docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,300 @@
+
+##### Egress
+Egress is not as simple as ingress. Currently, we support Locality Aware when atleast one client requires it, so we cannot distinguish between clients. Control-plane needs to send all dataplanes to the egress because there might be services sending requests to all zones. There are 2 options:
+- LoadBalacing based on metadata
+- Separate clusters
+- Support only Mesh scope for Egress
+
+Load balancing based on the metadata seems to be able to solve an issue but it has limitations. Because, egress knows all endpoints and does routing based on matching metadata the request might not obey the priority of zones. In this case we cannot configure different priorities for different clients.
+
+Separate cluster seems to be the only option that could solve this problem. When there is a different configuration of Load Balancing for the one of dataplane, control-plane sends to egress new a cluster `mesh-name:service-name:hash(tag + value)` with the routing based on the SNI: `service-name{mesh=mesh-name,hash=hash}`
+This cluster has a different configuration of priorities and only subset of zones. Dataplanes interested in this Load Balancing are going to have the same configuration of the cluster which sets SNI: `service-name{mesh=mesh-name,hash=hash}` and based in this SNI egress can make a routing decision to route to the specific cluster. That also requries each ingress to create a filter chain match for SNIs, to match an incoming request with the cluster.
+
+The last option doesn't create separate configuration for each client, but uses just Mesh scoped `MeshLoadBalancingStrategy` to configure egress. That reduces possibilites of configuring Egress but can be changed in the future.
+
+##### Ingress
+We don't configure ingresses.
+
+##### Serviceless
+On kubernetes you can create a deployment without the service, in this case control-planes creates default inbound so tags should be available.
+On universal you cannot create a dataplane without inbound or gateway definition.
+
+#### ClusterLoadBalancing cache
+Currently, while retrieving endpoints in the plugin, we are getting a pointer to the object in the cache that is shared between dataplanes. This won't work with localityAwareness load balancing because each dataplane might have a different subset of endpoints with different priorities. In the first iteration, we would like to try using it without CLA cache and measure the impact of the policy in the bigger cluster with many nodes. If the impact of not using the cache is significant, we might have to keep this configuration in the cache if there are more services on the same node. The cache might not improve the performance if there are instances of the same service in different localities (nodes, availability zones).
```

Comment:
How about saying:

- We're going to run tests with Mesh-Perf to see if the CLA cache has a real impact
- If results are good we'll remove the CLA cache
- Otherwise it's likely that the CLA cache will need to be rewritten to have different clusters depending on the dataplane tags.